src/nupic/algorithms/Connections.hpp

```diff
@@ -419,6 +420,45 @@ class Connections : public Serializable
                   CellIdx activePresynapticCell,
                   Permanence connectedPermanence) const;
 
+  /**
+   * The primary method in charge of learning.   Adapts the permanence values of
+   * the synapses based on the input SDR.  Learning is applied to a single
+   * segment.  Permanence values are increased for synapses connected to input
+   * bits that are turned on, and decreased for synapses connected to inputs
+   * bits that are turned off.
+   *
+   * @param segment  Index of segment to apply learning to.  Is returned by 
+   *        method getSegment.
+   * @param inputVector  An SDR
+   * @param increment  Change in permanence for synapses with active presynapses.
+   * @param decrement  Change in permanence for synapses with inactive presynapses.
+   */
+  void adaptSegment(Segment segment, SDR &inputs,
+                    Permanence increment,
+                    Permanence decrement);
+
+  /**
+   * Ensures a minimum number of connected synapses.  This raises permance
+   * values until the desired number of synapses have permanences above the
+   * permanenceThreshold.  This is applied to a single segment.
+   *
+   * @param segment  Index of segment on cell.   Is returned by method getSegment.
+   * @param permanenceThreshold  Connected threshold of synapses
+   * @param segmentThreshold  Desired number of connected synapses
+   */
+  void raisePermanencesToThreshold(Segment    segment,
+                                   Permanence permanenceThreshold,
+                                   UInt       segmentThreshold,
+                                   Permanence synPermBelowStimulusInc);
+
+  /**
+   * Modify all permanence on the given segment, uniformly.
+   *
+   * @param segment  Index of segment on cell. Is returned by method getSegment.
+   * @param delta  Change in permanence value
+   */
+  void bumpSegment(Segment segment, Permanence delta);
```

Comment:
these 3 methods above: 
This is what I wanted to discuss, what should belong to Connections, what should remain in SP (and use connections_;) ? 
Connections has for now been only used by TM, now will be shared by TM+SP. 
Seems data manipulation should be in Conn, specific functionality in SP, TM,..

Reply:
One of my design goals for the connections module is that data should enter and results leave. The user should not need to see the internal workings of it. It should hide permanences and synapse connected state, unless the user explicitly requests them via a getter method.  This way we have more room to optimize the internal workings of it.  
The user should request what they want (from the connections) instead of telling how they want to modify (the connections).

Reply:
I basically agree, 

> data should enter and results leave 

but what I'm saying is what is the line? You could say that it should provide compute() method (=basically rewrite SP in it)

> hide permanences and synapse connected state 

ok, this sounds reasonable

Reply:
Would it make sense to split the class per the use indicated above? Ie PotentialConnections, FeedforwardConnections? If it makes the logic and implementation simpler...just a thought for the future.

Reply:
I'm not sure I understand what you're saying.  

The connections class, although now a shared tool, is still going to have methods which have a single user, and I think this is a good thing.  The connections class is a good place to keep random utilities, which might someday be useful.  If connections class breaks into several variants, I worry that it will become more difficult to use.